instantout/reservation/manager.go

```diff
@@ -0,0 +1,260 @@
+				reservationRes.ReservationId)
+			err := m.newReservation(reservationRes)
+			if err != nil {
+				return err
+			}
+
+		case err := <-newBlockErrChan:
+			return err
+
+		case <-runCtx.Done():
+			log.Debugf("Stopping reservation manager")
+			return nil
+		}
+	}
+}
+
+// newReservation creates a new reservation from the reservation request.
+func (m *Manager) newReservation(
+	req *reservationrpc.ServerReservationNotification) error {
+
+	var reservationID ID
+	err := reservationID.FromByteSlice(
+		req.ReservationId,
+	)
+	if err != nil {
+		return err
+	}
+
+	serverKey, err := btcec.ParsePubKey(req.ServerKey)
+	if err != nil {
+		return err
+	}
+
+	// Create the reservation state machine.
+	reservationFSM := NewFSM(
+		m.runCtx, m.cfg,
+	)
+
+	m.Lock()
+	currentHeight := uint32(m.currentHeight)
+	m.Unlock()
+
+	initContext := &InitReservationContext{
+		reservationID: reservationID,
+		serverPubkey:  serverKey,
+		value:         btcutil.Amount(req.Value),
+		expiry:        req.Expiry,
+		heightHint:    currentHeight,
+	}
+
+	// Send the init event to the state machine.
+	err = reservationFSM.SendEvent(OnServerRequest, initContext)
+	if err != nil {
+		return err
+	}
+
+	// We'll now wait for the reservation to be in the state where it is
+	// waiting to be confirmed.
+	err = reservationFSM.DefaultObserver.WaitForState(
+		m.runCtx, time.Minute, WaitForConfirmation, time.Second,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RegisterReservationNotifications registers a new reservation notification
+// stream.
+func (m *Manager) RegisterReservationNotifications(
+	ctx context.Context, reservationChan chan *reservationrpc.
+		ServerReservationNotification) error {
+
+	// In order to create a valid lsat we first are going to call
+	// the FetchL402 method.
+	_, err := m.cfg.ReservationClient.FetchL402(
+		ctx, &reservationrpc.FetchL402Request{},
+	)
+	if err != nil {
+		return err
+	}
+
+	// We'll now subscribe to the reservation notifications.
+	reservationStream, err := m.cfg.ReservationClient.
+		ReservationNotificationStream(
+			ctx, &reservationrpc.ReservationNotificationRequest{},
+		)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+
+			default:
+				reservationRes, err := reservationStream.Recv()
+				if err != nil {
+					log.Errorf("Error receiving "+
+						"reservation: %v", err)
+					// If we encounter an error, we'll
+					// try to reconnect.
+					for {
+						select {
+						case <-ctx.Done():
+							return
+						default:
+							err = m.RegisterReservationNotifications(
+								ctx, reservationChan,
+							)
+							if err == nil {
+								log.Debugf(
+									"Successfully " +
+										"reconnected",
+								)
+								// If we were able to
+								// reconnect, we'll
+								// return.
+								return
+							}
+							log.Errorf("Error "+
+								"reconnecting: %v",
+								err)
+
+							time.Sleep(10 * time.Second)
```

Comment:
Instead of sleeping would be nicer to use `time.After` and (initialize it with zero so it fires immediately), that way terminating would be a non-issue.